File: src/data/courses/dsa/dynamic-programming/string-dp/longest-palindromic-subsequence/code/python/longest_palindromic_subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s) -> int:
        r=s[::-1]; n=len(s); dp=[0]*(n+1)
        for i in range(1,n+1):
            prev=0
            for j in range(1,n+1):
                t=dp[j]; dp[j]=prev+1 if s[i-1]==r[j-1] else max(dp[j],dp[j-1]); prev=t
        return dp[n]

File: src/data/courses/dsa/dynamic-programming/string-dp/longest-palindromic-subsequence/code/java/LongestPalindromicSubsequence.java
class Solution {
    public int longestPalindromeSubseq(String s) {
        String r=new StringBuilder(s).reverse().toString();
        int n=s.length(); int[] dp=new int[n+1];
        for(int i=1;i<=n;i++){int prev=0;
            for(int j=1;j<=n;j++){int t=dp[j];dp[j]=s.charAt(i-1)==r.charAt(j-1)?prev+1:Math.max(dp[j],dp[j-1]);prev=t;}}
        return dp[n];
    }
}